### DISCUSSION TITLE
How do I display an error for a server component?

### DISCUSSION THREAD

**Discussion by Tmcerlean** [vercel/next.js#53618] [Help]
### Summary

Here is the structure of a page (/app/(routes)/vehicles/page.jsx) in my app:

```
const Vehicles = ({ searchParams }) => {
    return (
        <Container>
            <ModalHeader />
            <Suspense fallback={<FiltersModalFallback initialMake={initialMake} vehicleType={vehicleType} />}>
                <FiltersModalData 
                    initialMake={initialMake}
                    make={make}
                    searchParams={searchParams}
                    vehicleType={vehicleType}
                />
            </Suspense>
            <ContentWrapper>
                <ErrorBoundary fallback={<div>Error</div>}>
                    <Suspense fallback={<LoadingListings />}>
                        <AdvertListingsData 
                            initialMake={initialMake}  
                            make={make}
                            searchParams={searchParams}
                            vehicleType={vehicleType}
                        />
                    </Suspense>
                </ErrorBoundary>
            </ContentWrapper>
        </Container>
)}
```

I am attempting to create an `ErrorBoundary` component which displays an error in place of the child server component `AdvertListingData` if the async db call within it fails. Unfortunately this does not seem to work. 

When there is an error in the async call the error.js file I created within the /vehicles/ directory (/app/(routes)/vehicles/error.jsx) triggers, but the `ErrorBoundary` component does not catch anything. The issue is that the error.js file replaces the entire vehicles page and not just the `AdvertListingsData` component. Is there any way around this?

For reference, here is the ErrorBoundary client component:

```
"use client"

import React from "react";


class ErrorBoundary extends React.Component {
    constructor(props) {
        super(props);
        this.state = { hasError: false };
    }

    static getDerivedStateFromError(error) {
        // Update state so the next render will show the fallback UI.
        return { hasError: true };
    }

    componentDidCatch(error, info) {
        // Example "componentStack":
        // in ComponentThatThrows (created by App)
        // in ErrorBoundary (created by App)
        // in div (created by App)
        // in App
        logErrorToMyService(error, info.componentStack);
    }

    render() {
        if (this.state.hasError) {
            // You can render any custom fallback UI
            return this.props.fallback;
        }

        return this.props.children;
    }
};

export default ErrorBoundary;
```

Thank you.

### Additional information

I was wondering if this is perhaps due to the error occuring on the server and so, given that the ErrorBoundary is a client component, it is not rendered before the error is triggered and so this is passed up to the parent (error.js) file?

**Comment by Tmcerlean** [comment]
After some more reading I found this in the legacy react docs here (https://legacy.reactjs.org/docs/error-boundaries.html#introducing-error-boundaries): 

> Note
> 
> Error boundaries do not catch errors for:
> 
> Event handlers ([learn more](https://legacy.reactjs.org/docs/error-boundaries.html#how-about-event-handlers))
> Asynchronous code (e.g. setTimeout or requestAnimationFrame callbacks)
> Server side rendering
> Errors thrown in the error boundary itself (rather than its children)

I assume this means that it is impossible to catch async errors in a server component. However, it somehow remains possible to be caught by the error.js file?

**Reply by jviall** [reply]
@Tmcerlean I'm in the exact same boat. I've reported an issue on this here: https://github.com/vercel/next.js/issues/58754. Have you found any work arounds for this? the `error.js` file does not provide the granularity of error UI that I need.
